### ISSUE TITLE
breaking change request: remove support for 64-bit (and unsigned 32-bit?) numbers

### ISSUE THREAD

**Issue by austinmilt** [webrpc/webrpc#283]
### Problem

While Go natively supports 64-bit numbers, not all languages do, including the biggest offender: [JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER). This can lead to unexpected behavior when numbers are truncated. Right now afaict, this behavior is implicit in the gen-typescript and gen-javascript repos. The reference server (webrpc-test) does not actually test this behavior, but that's another issue.

#### (current) Defer to each language and architecture
With this approach, webrpc makes no claim about how numbers should be standardized beyond the JSON spec such that e.g. `{"a": 18446744073709551615}` (which is valid JSON) is also valid in webrpc.

- some languages will balk when parsing an invalid number to a primitive type leading to unexpected errors for valid JSON
- some languages will silently fail by truncating numbers (e.g. JavaScript and Dart)
- individual webrpc-generators can partially solve this using special classes like `BigInt`, but that may have unintended side effects, especially when client and server are using different languages, and generally makes math operations more difficult

#### Encode all numbers as strings
With this approach, all numbers are encoded and decoded as strings, such that you end up with, e.g. `{"a": "18446744073709551615"}`.

- makes it clear that there's an extra step for treating all numbers, and that webrpc-users must think about how to get numbers in and out of that format, including accepting features like language-level truncation
- reduces the readability of the JSON, since it's unclear if the string is actually a number or not
- breaks built-in JSON utilities that will automatically parse the value as a string

#### Remove support for 64-bit numbers (for now)
With this approach, generated code will be expected to error when a number outside the accepted range is used, such that, e.g. `{"a": 18446744073709551615}` would result in an error. If doing this, we should also consider removing support for `uint32` since this may overflow in situations where only 32-bit signed integers are available.

- caters to the most restrictive cases where webrpc is likely to be used
- leaves no ambiguity about how big numbers should be handled
- reduces the number of use cases for webrpc, including easy adoption into systems that already have strict requirements for big numbers, though, in that scenario...
- ...developers can still choose to using strings to represent their numbers and write special second-layer decoders

### Recommendation
[Big numbers are hard to support.](https://github.com/protocolbuffers/protobuf/issues/16270). It would be better to make this wide-spread issue explicit for users of webrpc than to potentially fail silently. My suggestion is to kick the can down the road by removing support for big numbers (at least 64-bit numbers) and make a standard decision when users of webrpc make the demand.

**Comment by VojtechVitek** [comment]
Referencing https://github.com/webrpc/webrpc/issues/143.

**Comment by VojtechVitek** [comment]
@austinmilt Is there a primitive in Dart similar to JavaScript's BigInt? There's a proposal to use it at https://github.com/webrpc/webrpc/issues/143.

**Comment by austinmilt** [comment]
@VojtechVitek Yes Dart does have `BigInt`, but the built-in JSON parser in Dart uses Javascript's `JSON.parse` when deployed on the web, and this leads to the number being truncated before there's an opportunity to do anything about it "in Dart". I opened [an issue about this with the Dart core library](https://github.com/dart-lang/sdk/issues/55499) and we'll see where it goes, but for the moment it's a bit of chicken-and-egg. Properly supporting parsing to `BigInt` would mean an additional dependency or writing a custom JSON parser, both of which are acceptable if not ideal.
